misk/src/main/kotlin/misk/web/actions/ReadinessCheckAction.kt

```diff
@@ -37,20 +47,42 @@ class ReadinessCheckAction @Inject internal constructor(
       // Don't do healthchecks if services haven't all started. The app isn't in a good state yet,
       // and a health check could end up triggering random errors that we don't want to flood the
       // logs with.
-      return Response("", statusCode = 503)
+      return@runBlocking Response("", statusCode = 503)
+    }
+
+    // First run needs to block and get statuses
+    if (lastUpdate == null) {
+      refreshStatuses()
+    } else {
+      // This needs to be independent of the max age check so that services can recover
+      // If refreshStatuses() is never run then a failed/stale check can never recover
+      if (clock.instant().isAfter(lastUpdate!!.plusMillis(config.readiness_refresh_interval_ms.toLong()))) {
+        // Launch in the background so that we can still return the cached state
+        launch(Job()) {
+          refreshStatuses()
+        }
+      }
+
+      if (clock.instant().isAfter(lastUpdate!!.plusMillis(config.readiness_max_age_ms.toLong()))) {
+        logger.info("Failed health check: last status check is older than max age")
+        return@runBlocking Response("", statusCode = 503)
+      }
     }
 
-    val failedHealthChecks = healthChecks
-      .map { it.status() }
-      .filter { !it.isHealthy }
+    val failedHealthChecks = statuses.filter { !it.isHealthy }
 
     if (failedHealthChecks.isEmpty()) {
-      return Response("", statusCode = 200)
+      return@runBlocking Response("", statusCode = 200)
     }
 
     for (healthCheck in failedHealthChecks) {
       logger.info("Failed health check: ${healthCheck.messages}")
     }
-    return Response("", statusCode = 503)
+    return@runBlocking Response("", statusCode = 503)
+  }
+
+  private fun refreshStatuses() {
+    statuses = healthChecks.map { it.status() }
```

Comment:
Should we add a synchronization mechanism around this state? We should normally never have concurrent access, but if a service has slow health checks or a frequent readiness probe it could happen